src/libcrypto/poly_math.c

```diff
@@ -1,39 +1,254 @@
+    }
+}
+
+void double_copy(field_double* from, field_double* to) {
+    element* fromptr = from->e;
+    element* toptr = to->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(from->e); i++) {
+        *toptr++ = *fromptr++;
+    }
+}
+
+void single_to_double(field_2n* from, field_double* to) {
+    short i;
+
+    double_null(to);
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(from->e); i++) {
+        to->e[i + ARRAY_COUNT(to->e) - ARRAY_COUNT(from->e)] = from->e[i];
+    }
+}
+
+void double_to_single(field_double* from, field_2n* to) {
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(to->e); i++) {
+        to->e[i] = from->e[i + ARRAY_COUNT(from->e) - ARRAY_COUNT(to->e)];
+    }
+}
+
+void multiply_shift(field_double* a) {
+    element* eptr = &a->e[ARRAY_COUNT(a->e) - 1];
+    element bit = 0;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++, eptr--) {
+        element temp = *eptr;
+        *eptr = (temp << 1) | bit;
+        bit = temp >> 31;
+    }
+}
 
 INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", poly_mul_partial);
 
 INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", multiply_shift_n);
 
 INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", divide_shift_n);
 
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", extract_masked_bits);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", zero_masked_bits);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", shift_and_add);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", poly_mul);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", cus_times_u_to_n);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", is_less_than);
+void extract_masked_bits(field_double* a, field_double* mask, field_double* result) {
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        result->e[i] = a->e[i] & mask->e[i];
+    }
+}
+
+void zero_masked_bits(field_double* a, field_double* mask) {
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        a->e[i] &= ~mask->e[i];
+    }
+}
+
+void shift_and_add(field_double* temp, field_double* extract_mask) {
+    field_double temp1;
+    field_double temp_masked;
+
+    extract_masked_bits(temp, extract_mask, &temp_masked);
+    zero_masked_bits(temp, extract_mask);
+    divide_shift_n(&temp_masked, 159);
+    double_add(temp, &temp_masked, &temp1);
+    divide_shift_n(&temp_masked, 74);
+    double_add(&temp1, &temp_masked, temp);
+}
+
+void poly_mul(field_2n* a, field_2n* b, field_2n* c) {
+    field_double temp;
+    field_double extract_mask;
+
+    double_null(&extract_mask);
+    extract_mask.e[0] = 0xFFFFF;
+    extract_mask.e[1] = 0xFFFFFFFF;
+    extract_mask.e[2] = 0xFFFFFFFF;
+    extract_mask.e[3] = 0xFFFFFFFF;
+    extract_mask.e[4] = 0xFFFFFFFF;
+    extract_mask.e[5] = 0xFFE00000;
+    poly_mul_partial(a, b, &temp);
+    shift_and_add(&temp, &extract_mask);
+    divide_shift_n(&extract_mask, 159);
+    extract_mask.e[7] = -0x200;
```

Comment:
```suggestion
    extract_mask.e[7] = 0xFFFFFE00;
```
I think this makes slightly more sense?